http/README.md

```diff
@@ -0,0 +1,133 @@
+
+- Browsers communicating with other libp2p nodes without needing a WebSocket, WebTransport, or WebRTC connection.
+- HTTP only edge workers can run application protocols and respond to peers on the network.
+- `curl` from the command line can make requests to other libp2p nodes.
+
+The HTTP transport will also allow application protocols to make use of HTTP intermediaries such as HTTP caching, and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:
+
+- Connectivity – Work on top of WebRTC, WebTransport, QUIC, TCP, or an HTTP transport.
+- Hole punching – Work with peers behind NATs.
+- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
+- Peer discovery – Learn about a peer given their peer id.
+
+## HTTP Semantics vs Encodings vs Transport
+
+HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.
+
+
+```mermaid
+graph TB
+    subgraph "HTTP Semantics"
+        HTTP
+    end
+    subgraph "Encoding"
+        HTTP1.1[HTTP/1.1]
+        HTTP2[HTTP/2]
+        HTTP3[HTTP/3]
+    end
+    subgraph "Transports"
+        Libp2p[libp2p streams]
+        HTTPTransport[HTTP transport]
+    end
+    HTTP --- HTTP1.1
+    HTTP --- HTTP1.1
+    HTTP1.1 --- Libp2p
+    HTTP --- HTTP2
+    HTTP --- HTTP3
+    HTTP1.1 --- HTTPTransport
+    HTTP2 --- HTTPTransport
+    HTTP3 --- HTTPTransport
+```
+
+- *HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is
+  the stateless application-level protocol that you work with when writing HTTP
+  apis (for example).
+
+- *HTTP encoding* is the thing that takes your high level request/response
+  defined in terms of HTTP semantics and encodes it into a form that can be sent
+  over the wire.
+
+- *HTTP transport* is the thing that takes your encoded reqeust/response and
+  sends it over the wire. For HTTP/1.1 and HTTP/2, this is a TCP+TLS connection.
+  For HTTP/3, this is a QUIC connection.
+
+When this document says *HTTP* it is generally referring to *HTTP semantics*.
+
+## Interoperability with existing HTTP systems
+
+A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in
+their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
+component spec](https://github.com/libp2p/specs/pull/550) for more details.
+
+## Namespace
+
+libp2p does not squat the global namespace. libp2p application protocols can be
+discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615)
+`.well-known/libp2p`. This allows server operators to dynamically change the
```

Comment:
We use well-known resources [elsewhere](https://docs.libp2p.io/concepts/transports/webtransport/#webtransport-in-libp2p), e.g. `.well-known/libp2p-webtransport`.  Perhaps we should namespace this one too?

```suggestion
`.well-known/libp2p-http`. This allows server operators to dynamically change the
```

Reply:
Aesthetically it's nicer if we aren't referencing HTTP multiple times. `.well-known/libp2p` is a HTTP resource. It's kind of like saying "ATM machine". 

Technically this doesn't matter, but my vote is for `.well-known/libp2p`

Reply:
I'm not sure I agree. Given we have other well-known resources, `.well-known/libp2p` is ambiguous, `.well-known/libp2p-http` is not.

Though yes, from a technical perspective it just has to be a predictable string with a low chance of collision.

Reply:
@lidel can you be our tie breaker? `.well-known/libp2p` vs `.well-known/libp2p-http`

Reply:
Squatting `.well-known/libp2p` for a file may cause us problems in the future if we need to add something else.

I think we should go with either `.well-known/libp2p-http` ..or make a directory and put protocol mapping configuration in  `.well-known/libp2p/http`.

The latter (`libp2p/http`) feels a bit better to me, as it creates libp2p-specific directory/namespace, which is easier to map via reverse proxies, but other than that it is mostly aesthetics.

Reply:
Just a note that this is a map, so we could always add more things in a backwards compatible way.

I like namespacing with `.well-know/libp2p`. This lets us get a single IANA URI suffix if we want while still giving us flexibility in what goes under it.

I'll make the change to `.well-known/libp2p/protocols`, thanks. Mark this one as another success for IPFS' Chief Renaming Officer.

Reply:
Just to point out that we use [.well-known/libp2p-webtransport](https://docs.libp2p.io/concepts/transports/webtransport/#webtransport-in-libp2p) for WebTransport things already, `.well-known/libp2p-http` would be consistent with that, `.well-known/libp2p/protocols` looks like something different.

Reply:
This is different though. This is metadata about a peer's supported protocols. The well-know webtransport URI is about where to send the HTTP CONNECT request to.

Reply:
I'm not sure it is, I think how you interact with the well-known resource is a detail of the protocol that's unrelated to the path it uses.

Reply:
The `.well-known/libp2p/` suffix has the benefit of being a single thing we would need to register. Future versions of libp2p-webtransport may be placed under that suffix. Remember WebTransport isn't even out of draft status yet. We'll probably need to make a new multiaddr for webtransport-v1 just like we did with QUIC. Any other well-known resource we would want could also fit under that suffix.

Reply:
Changing this from "/.well-known/libp2p" to "/.well-known/libp2p/protocols" will break the ability to communicate with existing servers deployed without upgrading them first. Since these servers are running in places not under our control, they will not be immediately upgradable.

Because there are already deployments that use the old value, I suggest keeping it as `/.well-known/libp2p`

Reply:
I understand the pain in this rollout. Especially when the full rollout is outside your control. Maybe go-libp2p can provide a transition period to help out?

Since this spec itself is in draft we need flexibility in how things work. As much as I prefer the `.well-known/libp2p`, I don't think the argument of "we should use it because we've deployed this draft" is a strong one. Deployments of drafts are extremely useful to get some experience, but we shouldn't ossify to those decisions because we already made them.

Once this is merged, things will be stable and we won't break existing users. That's the guarantee that comes with merging this. And partly the reason why I haven't merged it yet. I want to build the js-libp2p side before merging and make sure that the interop works and is reasonable. I'm pretty close on that. Hoping to get it done by the end of the month :)